### PR TITLE
Upload code coverage

### PR DESCRIPTION
**Pull request by julianrex**
PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-ios/blob/main/CONTRIBUTING.md#contributor-license-agreement).

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

## Pull request checklist:
 - [X] Briefly describe the changes in this PR.
 - [X] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] ~Write tests for all new functionality. If tests were not written, please explain why.~
 - [ ] ~Add example if relevant.~
 - [ ] ~Document any changes to public APIs.~
 - [X] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [ ] ~Add an entry inside this element for inclusion in the `mapbox-maps-ios` changelog: `<changelog></changelog>`.~

### Summary of changes

This PR:

- Generates lcov data from the original `profraw` file created by simulator and device testing.
- Generates lcov JSON and parses into a custom coverage format. This is uploaded for internal consumption.
- Reports device test failures on merge-to-main (untested)
- Saves separate artifacts for each simulator test run

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -332,7 +361,6 @@ jobs:
                 message: '<$CIRCLE_BUILD_URL|Release build for \`$VERSION\` started.>'
                 include_visit_job_action: true
             - make-deps
-            - run: pip3 install awscli
```
Comment (neelmistry94) [inline .circleci/config.yml:335]: This is needed for s3 uploads

File: .circleci/config.yml [hunk 2]
```diff
@@ -375,7 +403,6 @@ jobs:
                 include_visit_job_action: true
             - checkout
             - make-deps
-            - run: pip3 install awscli
```
Comment (neelmistry94) [inline .circleci/config.yml:378]: I believe this was needed for making api-docs. There was a reason I had to add this in

File: .gitignore [hunk 3]
```diff
@@ -8,3 +8,4 @@ developer.xcconfig
 Carthage
 api-docs
 MapboxAccessToken
+.swiftpm/
```
Comment (neelmistry94) [inline .gitignore:11]: @macdrevx I believe this should not be in `.gitignore` right?

File: cloudformation/ci.template.js [hunk 4]
```diff
@@ -70,6 +70,20 @@ AWSTemplateFormatVersion: '2010-09-09',
...
+              Statement: [
+                {
+                  Action: ['s3:PutObject'],
+                  Effect: 'Allow',
+                  Resource: ['arn:aws:s3:::mapbox-loading-dock/raw/mobile_staging.codecoverage_v2/*',
+                             'arn:aws:s3:::mapbox-loading-dock/raw/mobile.codecoverage/*',
+                             'arn:aws:s3:::mapbox-loading-dock/raw/mobile_staging.codecoverage/*']
+                }
+              ]
+            }
```
Comment (neelmistry94) [inline cloudformation/ci.template.js:86]: did you already run the formation update job for this?

Peng just opened [this](https://github.com/mapbox/mapbox-maps-internal/issues/782) ticket for moving cloud formation to private repo. Maybe we can take this PR as an opportunity to JFDI

**Review by neelmistry94 — CHANGES_REQUESTED** [review]
Left comments inline to reflect changes requested

File: .circleci/config.yml (hunk 1 above)
Comment (julianrex) [reply .circleci/config.yml:335]: This has moved into the `make-deps` step.

File: .circleci/config.yml (hunk 2 above)
Comment (julianrex) [reply .circleci/config.yml:378]: Same comment as above. We may want to stream-line these.

File: cloudformation/ci.template.js (hunk 4 above)
Comment (julianrex) [reply cloudformation/ci.template.js:86]: Yep, update job is run. 

Let's address moving cloud formation later; it'll be a sizable job and affects multiple steps.

**Review by neelmistry94 — APPROVED** [review]

**Pull request merged**